uqcsbot/starboard.py

```diff
@@ -1,244 +1,598 @@
-            
-            self.base_blocked_messages += [recv_message.id]
-            Timer(self.ratelimit, self._rm_base_ratelimit, [recv_message.id]).start()
-        elif new_reaction_count > self.base_threshold and \
-                sb_message_id is not None and recv_message.id not in self.big_blocked_messages:
-            old_sb_message = await self.starboard_channel.fetch_message(sb_message_id)
-
-            await old_sb_message.edit(
-                content=f"{str(self.starboard_emoji)} {new_reaction_count} | {recv_message.channel.mention}"
+            # recieved_msg isn't None and we just sent the sb message, so it also shouldn't be None
+            self._starboard_db_add(
+                recieved_msg.id, recieved_msg.channel.id, new_sb_message.id
             )
 
-            if new_reaction_count >= self.big_threshold and not old_sb_message.pinned:
-                await old_sb_message.pin(reason=f"Reached {self.big_threshold} starboard reactions")
-            
-            self.big_blocked_messages += [recv_message.id]
-            Timer(self.ratelimit, self._rm_big_ratelimit, [recv_message.id]).start()
+            # start the base ratelimit
+            self.base_blocked_messages += [recieved_msg.id]
+            Timer(self.ratelimit, self._rm_base_ratelimit, [recieved_msg.id]).start()
+        elif reaction_count >= self.base_threshold and starboard_msg is not None:
+            # Above threshold, existing message? update it.
+            if (
+                reaction_count >= self.big_threshold
+                and starboard_msg.id not in self.big_blocked_messages
+            ):
+                await starboard_msg.pin(
+                    reason=f"Reached {self.big_threshold} starboard reactions."
+                )
+
+                # start the pin ratelimit
+                self.big_blocked_messages += [starboard_msg.id]
+                Timer(
+                    self.ratelimit, self._rm_big_ratelimit, [starboard_msg.id]
+                ).start()
+            elif (
+                starboard_msg.pinned
+                and starboard_msg.id not in self.big_blocked_messages
+            ):
+                await starboard_msg.unpin(
+                    reason=f"Fell below starboard threshold ({self.big_threshold})."
+                )
+
+            await starboard_msg.edit(
+                content=self._generate_message_text(reaction_count, recieved_msg)
+            )
+        else:
+            # Below threshold, or blocked from sending. Might need to delete.
+            if starboard_msg is not None:
+                self._starboard_db_remove(recieved_msg, starboard_msg)
+                await starboard_msg.delete()
 
+    """
+    The four handlers below here are more or less identical.
+    The differences are just that `add` deletes :starhaj:'s that come from the author of the message,
+    and `clear` doesn't need to check that the payload references :starhaj:.    
+    """
```

Comment:
Is a handler for deleting messages also needed, or is this a subclass of another handler?

Reply:
I might just add a wiki page with details about the starboard (because the list of different behaviours it has is getting quite extensive), and make the wiki links more obvious in the readme as a "START HERE" type thing.